galley/pkg/mcp/snapshot/snapshot.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshot
+
+import (
+	"sync"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+
+	mcp "istio.io/api/config/mcp/v1alpha1"
+	"istio.io/fortio/log"
+	"istio.io/istio/galley/pkg/mcp/server"
+)
+
+// Snapshot provides an immutable view of versioned resources.
+type Snapshot interface {
+	Resources(typ string) []*mcp.Envelope
+	Version(typ string) string
+}
+
+// Cache is a snapshot-based cache that maintains a single versioned
+// snapshot of responses per node. Cache consistently replies with the
+// latest snapshot.
+type Cache struct {
+	mu         sync.RWMutex
+	snapshots  map[string]Snapshot
+	status     map[string]*statusInfo
+	watchCount int64
+}
+
+// New creates a new cache of resource snapshots.
+func New() *Cache {
+	return &Cache{
+		snapshots: make(map[string]Snapshot),
+		status:    make(map[string]*statusInfo),
+	}
+}
+
+var _ server.Watcher = &Cache{}
+
+type responseWatch struct {
+	request   *xdsapi.DiscoveryRequest // original request
+	responseC chan<- *server.WatchResponse
+}
+
+type statusInfo struct {
+	mu                   sync.Mutex
+	node                 *core.Node
+	lastWatchRequestTime time.Time
+	watches              map[int64]*responseWatch
+}
+
+// Watch returns a watch for an MCP request.
+func (c *Cache) Watch(request *xdsapi.DiscoveryRequest, responseC chan<- *server.WatchResponse) (*server.WatchResponse, server.CancelWatchFunc) { // nolint: lll
+	nodeID := request.Node.GetId()
```

Comment:
Should this log a warning if the request comes with resource names?
Similarly, how about status?

Reply:
Added a warning if resource names (hints) are provided. 

ErrorDetails is handled by the lower-level grpc server.